<launch>
  <arg name="param_file" default="$(find-pkg-share lidar_marker_localizer)/config/lidar_marker_localizer.param.yaml"/>

  <!-- Topic names -->
  <arg name="input_lanelet2_map" default="~/input/lanelet2_map"/>
  <arg name="input_ekf_pose" default="~/input/ekf_pose"/>
  <arg name="input_pointcloud" default="~/input/pointcloud"/>

  <arg name="output_pose_with_covariance" default="~/output/pose_with_covariance"/>

  <arg name="service_trigger_node_srv" default="~/service/trigger_node_srv"/>

  <!-- container -->
  <arg name="pointcloud_container_name" default="/pointcloud_container" description="container name"/>

  <!-- whether use intra-process -->
  <arg name="use_intra_process" default="true" description="use ROS 2 component container communication"/>


  <load_composable_node target="$(var pointcloud_container_name)">
    <composable_node pkg="pointcloud_preprocessor" plugin="pointcloud_preprocessor::CropBoxFilterComponent" name="crop_box_filter_measurement_range">
      <param name="input_frame" value="base_link"/>
      <param name="output_frame" value="base_link"/>
      <param name="min_x" value="-10.0"/>
      <param name="max_x" value="10.0"/>
      <param name="min_y" value="0.0"/>
      <param name="max_y" value="7.5"/>
      <param name="min_z" value="-5.0"/>
      <param name="max_z" value="5.0"/>
      <param name="negative" value="False"/>
      <remap from="input" to="$(var input_pointcloud)"/>
      <remap from="output" to="measurement_range/pointcloud_ex"/>
      <extra_arg name="use_intra_process_comms" value="$(var use_intra_process)"/>
    </composable_node>

    <composable_node pkg="pointcloud_preprocessor" plugin="pointcloud_preprocessor::PassThroughFilterUInt16Component" name="ring_filter">
      <param name="input_frame" value="base_link"/>
      <param name="output_frame" value="base_link"/>
      <param name="filter_field_name" value="ring"/>
      <param name="filter_limit_min" value="5"/>
      <param name="filter_limit_max" value="45"/>
      <param name="filter_limit_negative" value="False"/>
      <param name="keep_organized" value="False"/>
      <remap from="input" to="measurement_range/pointcloud_ex"/>
      <remap from="output" to="ring_filter/pointcloud_ex"/>
      <extra_arg name="use_intra_process_comms" value="$(var use_intra_process)"/>
    </composable_node>

  </load_composable_node>

  <node pkg="lidar_marker_localizer" exec="lidar_marker_localizer" name="lidar_marker_localizer" output="log">
    <remap from="~/input/pointcloud_ex" to="ring_filter/pointcloud_ex"/>
    <remap from="~/input/ekf_pose" to="$(var input_ekf_pose)"/>
    <remap from="~/input/lanelet2_map" to="$(var input_lanelet2_map)"/>

    <remap from="~/output/pose_with_covariance" to="$(var output_pose_with_covariance)"/>

    <remap from="~/service/trigger_node_srv" to="$(var service_trigger_node_srv)"/>

    <param from="$(var param_file)"/>
  </node>

</launch>
